Enables service installation with a name

laardee opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

This PR enables to rename the service when installing from GitHub, for example `serverless install --name new-service-name --url https://github.com/SC5/serverless-messenger-boilerplate` will install `serverless-messenger-boilerplate` to folder `new-service-name` and change `service` value in `serverless.yml` to `new-service-name`.

Closes https://github.com/serverless/serverless/issues/2626

## How did you implement it:

Added new parameter `name` to install class and renameService helper function for renaming service in yaml file.

## How can we verify it:

`serverless install --name new-service-name --url https://github.com/SC5/serverless-messenger-boilerplate` should install service to `new-service-name` folder and change service name to `new-service-name`.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below


***Is this ready for review?:*** YES

laardee commented:
About tests, what kind of a test should I write for this? The current installation test just seems (?) to mock download or something like that (the url is invalid).

pmuens commented:
Hey @laardee. Great addition to the `install` command! I like it 👍  💯 

Hmm. Couldn't you stub the `download` again and return something there (pretending the download succeeds). And after that test if the `renameService` function is called.
Furthermore you can unit test the `renameService` in isolation.

Maybe @eahefnawy and @flomotlik can add some thoughts here.

Today I stumbled upon an improvement for the `proxyquire` usage. I've implemented it here:
https://github.com/serverless/serverless/blob/master/lib/classes/Utils.test.js#L21-L29

Basically it should be moved in the `beforeEach()` so that you get a fresh stub every time.

Let me know if you need any help!

laardee commented:
@pmuens thanks, I wrote unit test for renameService https://github.com/serverless/serverless/pull/2616/commits/af53bcb9d7dd6421ef0d01daaf1eeeb2106962cf, it creates serverless.yml and package.json files (only service and name keys), then it uses renameService to change those and verifies that. 

I also moved downloadStub to beforeEach as you suggested and added one expect to check that the path that is created contains the new name for service https://github.com/serverless/serverless/pull/2616/commits/df499e72786f02594b6dfae3099ef7598a91c35c.

pmuens commented:
Hey @laardee thank you. Sounds good! 👍 

Will review and take a deep dive ASAP.

laardee commented:
I changed yml read/write in renameService function to use plain fs module with string replace to preserve comments in serverless.yml, js-yaml module doesn't seem to support comments https://github.com/nodeca/js-yaml/issues/305.

pmuens commented on docs/providers/aws/cli-reference/install.md:
````diff
@@ -32,4 +33,12 @@ serverless install --url https://github.com/some/service
+serverless install --url https://github.com/johndoe/authentication --name my-authentication
+```
+
+This example will download the .zip file of the `authentication` service from GitHub, create a new directory with the name `my-authentication` in the current working directory, unzips the files in this directory and renames service to `my-authentication` if `serverless.yml` exists in the service root.
````
"... and renames the service"

laardee replied:
fixed

pmuens commented on lib/plugins/install/install.js:
```diff
@@ -22,6 +24,11 @@ class Install {
           },
+          name: {
+            usage: 'Name for the service',
+            required: false,
```
This could be left out as it means that same as `require: false` (AFAIK we don't follow this pattern in other plugins).

laardee replied:
fixed

pmuens commented on lib/plugins/install/install.js:
```diff
@@ -30,6 +37,30 @@ class Install {
+          .then(contents => this.serverless.utils.writeFile(packageFile, contents)),
+      ])
+        .then(() => resolve(` as "${name}"`))
+        .catch((error) => {
```
This could be enhanced following this pattern we use throughout the codebase (https://github.com/serverless/serverless/blob/5430fd87344595e67680f68fc1b3e594a5084f17/lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js#L43-L50).

pmuens replied:
Or you could let Serverless catch the error and remove the `catch` block altogether.

laardee replied:
I refactored this to be synchronous and used `this.serverless.classes.Error` for throwing error. Error message could still be better.

pmuens commented on lib/plugins/install/tests/install.js:
```diff
@@ -77,5 +129,18 @@ describe('Install', () => {
       });
     });
+
+    it('should download and rename the service based on the GitHub URL', () => {
```
Can we enhance this test somehow that the call to the renaming method is tested as well?

laardee replied:
@pmuens, so basically either serverless.yml should be created to the download folder beforehand or downloadStub should pass zip containing serverless.yml as a payload. The first one might not work, because file should be created to folder before running the install function, which checks if the folder exists and if so it won't install the package. I'll check if I can pass the zip as a payload.

laardee replied:
Now the downloadStub mocks the zip loading and saves serverless.yml, then installation process renames the service value and finally the result is tested.

pmuens commented on lib/plugins/install/install.js:
```diff
@@ -30,6 +37,30 @@ class Install {
+        readFile(serviceFile, 'utf-8')
+          .then(contents =>
+            contents.replace(/service\s*:.+/gi, (match) => {
+              const fractions = match.split('#');
```
Are commented out parts in a `serverless.yml` file that common?

laardee replied:
I don't know about others, but at least we have some comments on our workshop boilerplates and other similar project templates.

pmuens commented on lib/plugins/install/install.js:
```diff
@@ -30,6 +37,30 @@ class Install {
+              return fractions.join(' #');
+            }))
+          .then(contents => writeFile(serviceFile, contents)),
+        this.serverless.utils.readFile(packageFile)
```
Does a service always have a `package.json` file?

laardee replied:
no, fixed.

laardee commented:
@pmuens, ready for review round two :sweat_smile:

pmuens commented:
Hey @laardee 
Thank you very much for the updates to this PR! 👍 
I followed this PR the other days and it looks really promising!

I take a deep dive later again because we're currently focused on 1.2 (https://github.com/serverless/serverless/milestone/16) to get this out of the door next week.

Again. Thanks for working on this really good addition! 👍  💃

laardee commented:
@pmuens, have you already had time to look into this? I know this is not the first priority PR so I don't mind waiting and rebasing 😄

pmuens commented:
@laardee thanks for getting back. 👍 

Unfortunately I didn't have any time yet 😞 . But we'll plan the next milestones soon and will prioritize this PR soon.

laardee commented:
@pmuens, I fixed the code to work with https://github.com/serverless/serverless/pull/2816. Unit tests are passing but the lint fails because of the metrics plugin. Is someone fixing that already?

laardee commented:
rebased, tests are now passing 👌

laardee commented:
I wonder if 🎅 comes before next review round. 😀
